### PR TITLE
Adding pluck field functionality to extractors

### PR DESCRIPTION
**Pull request by ramanakumars**
Created a new function that can add keywords to output dictionary from extractors, with values from the classification metadata. 

cc @CKrawczyk

### CODE REVIEW

File: .idea/.gitignore [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Default ignored files
+/shelf/
+/workspace.xml
+# Datasource local storage ignored files
+/dataSources/
+/dataSources.local.xml
+# Editor-based HTTP Client requests
+/httpRequests/
```
Comment (CKrawczyk) [inline .idea/.gitignore:8]: Not exactly sure what this `.idea` folder is?  Can the entire folder be added to the main level `.gitignore` or does it serve some function?

File: panoptes_aggregation/extractors/pluck_extractor.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+from .extractor_wrapper import extractor_wrapper
+from .utilities import pluck_fields
+
+
+def slugify_or_null(s):
+    '''Slugify value while casting `null` as a string fisrt'''
+    if (s is None) or (isinstance(s, bool)) or (isinstance(s, int)):
+        return str(s)
+    else:
+        return slugify(s, separator='-')
```
Comment (CKrawczyk) [inline panoptes_aggregation/extractors/pluck_extractor.py:11]: Import this function from the question extractor rather than rewrite it:
```python
from .question_extractor import slugify_or_null
```

File: panoptes_aggregation/extractors/utilities.py [hunk 3]
```diff
@@ -11,3 +11,57 @@ def annotation_by_task(classification_in):
...
+    -----
+    classification : dict
+        A dictionary containing an `annotations` key that is a list of
+        panoptes annotations
+
+    pluck_keys : string
+        String with a list-like formatting that maps the output key to
+        the location of the value in the classification dict. Must begin
+        with a `[` and end with `]`, with each key map separated by a `,`.
+        Each entry should be in the form key:location (see examples below).
```
Comment (CKrawczyk) [inline panoptes_aggregation/extractors/utilities.py:35]: I am not convinced this is the best format for `pluck_keys`.  If this is instead passed in as a dictionary on the URL you can run: 

```python
import ast
pluck_key_dict = ast.literal_eval(pluck_keys)

for key, value in pluck_key_dict.items():
   ...
```

Less code and also matches the same way subtask information is passed into extractors and reducers for drawing tasks.

Note: My main concern is string parsing can get quite messy and lead to weird edge cases when badly formatted strings are passed into them.  `ast.literal_eval` will at some level protect from arbitrary code injection https://docs.python.org/3/library/ast.html#ast.literal_eval

File: panoptes_aggregation/extractors/pluck_extractor.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    extraction : dict
+        A dictionary with the entries defined by the `pluck` argument
+
+    See also
+    --------
+    .utilities.pluck_fields
+
+    '''
+    if pluck is not None:
+        return pluck_fields(classification, pluck)
```
Comment (CKrawczyk) [inline panoptes_aggregation/extractors/pluck_extractor.py:35]: Why is this its own extractor?  Wouldn't the `extractor_wrapper` also add the pluck fields resulting in them being added twice?  Might be more useful to make this a pass-through extractor that returns an empty dictionary and let the wrapper handle the plucking.

**Review by CKrawczyk — CHANGES_REQUESTED** [review]
Some general comments:

- Create a new branch when working on a PR rather than working directly on the master branch (even in your own fork), this just makes things more clear in the git history.
- You will need to create some unit tests for this code, see `panoptes_aggregation/extractor_tests/test_question_extractor.py` for some example tests.
- If this can be used with any extractor you will want at least one test that sets the pluck keyword for an existing extractor (the question extractor will likely be easiest to make a new test for).  This test could go in the existing test file for that extractor or in a new file, I don't really have a preference.

Once this extractor is done other work that will need to be done (in different PR):
- The reducer wrapper will need to be updated to remove all `pluck.*` fields from extracts and pass them along if the reducer is configured to accept them as a list of dictionaries (very similar to the `gold_standard` keyword that already exists).  This will make this data "opt-in" on a reducer-by-reducer level.
  - A reducer that requires the pluck fields would use something like `@reducer_wrapper(pluck=True)`
- Explicit tests should be added to at least one existing reducer to ensure the wrapper removes the extra fields correctly.

More specific comments on the code are placed inline.

File: panoptes_aggregation/extractors/pluck_extractor.py (hunk 4 above)
Comment (ramanakumars) [reply panoptes_aggregation/extractors/pluck_extractor.py]: I removed the pluck_extractor completely since it's functionality is now tied to the other extractors. Also removed the `.idea` folder and added it to `.gitignore` in case it pops up again. Also made the necessary changes to the `pluck_fields` function and added a unit tests for the question extractor. We're still looking at the other extractors and can maybe add those tests in a future PR?

File: .idea/.gitignore (hunk 1 above)
Comment (ramanakumars) [reply .idea/.gitignore:8]: Removed the `.idea` folder and added it to `.gitignore` in case it pops up again.

File: panoptes_aggregation/extractors/pluck_extractor.py (hunk 2 above)
Comment (ramanakumars) [reply panoptes_aggregation/extractors/pluck_extractor.py]: I removed the pluck_extractor completely since it's functionality is now tied to the other extractors.

File: panoptes_aggregation/extractors/utilities.py (hunk 3 above)
Comment (ramanakumars) [reply panoptes_aggregation/extractors/utilities.py]: I made the necessary changes to the `pluck_fields` function

**Comment by ramanakumars** [comment]
I've added a unit test for the question extractor. We're still looking at the other extractors and can maybe add those tests in a future PR?

File: panoptes_aggregation/extractors/utilities.py (hunk 3 above)
Comment (CKrawczyk) [reply panoptes_aggregation/extractors/utilities.py]: 🤦 I edited this comment but forgot to hit save before submitting the review. After I first wrote this I thought of a better way to do this that is more in line with the other code in the repo.  I will put some inline comments in the new code.

File: panoptes_aggregation/extractors/utilities.py [hunk 5]
```diff
@@ -11,3 +12,57 @@ def annotation_by_task(classification_in):
...
+              "is_gold_standard": "False" \
+            }, \
+        }}
+    >>> pluck_keys = '{"gold_standard":"subject.metadata.is_gold_standard", "true_value":"subject.metadata.uber_flag_digit"}'
+    >>> pluck_fields(classification, pluck_keys)
+    defaultdict(<class 'list'>, {'pluck.gold_standard': 'False', 'pluck.true_value': '4'})
+    '''
+    answers = {}
+
+    pluck_key_dict = ast.literal_eval(pluck_keys)
```
Comment (CKrawczyk) [inline panoptes_aggregation/extractors/utilities.py:55]: To bring this more in line with the other extractor code (e.g. subtask details) can this evaluation be moved to the `extractor_wrapper` on line 27 and this code changed to take in a dictionary object rather than a string formatted like a dictionary?

When not sending the object of a URL it is much easier to just send in a dictionary object (e.g. the configuration yaml files create python dictionaries anyways, might as well keep it in that format), the only time it needs conversion from a string is if submitted via URL and that is exactly what the `extractor_wrapper` is designed to handle.

File: panoptes_aggregation/tests/extractor_tests/test_question_extractor.py [hunk 6]
```diff
@@ -87,3 +87,29 @@
     kwargs={'task': 'T1'},
     test_name='TestSingleTwoTasksKey'
 )
+
+single_pluck_classification = {"id": 359841171,
```
Comment (CKrawczyk) [inline panoptes_aggregation/tests/extractor_tests/test_question_extractor.py:91]: Minor style change, put everything after `{` on the next line with an indent.

File: panoptes_aggregation/tests/extractor_tests/test_question_extractor.py [hunk 7]
```diff
@@ -87,3 +87,29 @@
...
+        "id": 68561182,
+        "metadata": {
+            "filename": "uds_H.gf19767.png",
+            "uber_flag_digit": "4",
+            "is_gold_standard": "False"
+        }
+    }
+}
+
+single_pluck_keys = '{"gold_standard": "subject.metadata.is_gold_standard", "true_value": "subject.metadata.uber_flag_digit"}'
```
Comment (CKrawczyk) [inline panoptes_aggregation/tests/extractor_tests/test_question_extractor.py:106]: As noted above, this should be a dictionary rather than a string at this point.  Minor note, to keep the tests a bit more readable (and future git-diffs easier to see changes) I tend to space all dicts out to one item per line:

```python
single_pluck_keys = {
    "gold_standard": "subject.metadata.is_gold_standard",
    "true_value": "subject.metadata.uber_flag_digit"
}
```

File: panoptes_aggregation/tests/extractor_tests/test_question_extractor.py [hunk 8]
```diff
@@ -87,3 +87,29 @@
...
+
+single_pluck_expected = {"0": 1, "pluck.gold_standard": "False", "pluck.true_value": "4"}
```
Comment (CKrawczyk) [inline panoptes_aggregation/tests/extractor_tests/test_question_extractor.py:108]: Minor style change, space this out one item per line.

**Review by CKrawczyk — CHANGES_REQUESTED** [review]
A few more comments (I thought I included these comments yesterday but I forgot to hit the save button on GitHub 🤦).

**Comment by CKrawczyk** [comment]
Also, feel free to add tests to other extractors, but seeing as it is baked into the `extractor_wrapper` code that they all use I don't think it is strictly necessary.  If it works for one of them it should work for all of them (unless you can think of an interesting edge case).

File: panoptes_aggregation/tests/extractor_tests/test_question_extractor.py (hunk 7 above)
Comment (ramanakumars) [reply panoptes_aggregation/tests/extractor_tests/test_question_extractor.py:106]: I might be misunderstanding your point, but wouldn't it still be a string at this point, because the `pluck` input parameter will be evaluated as a string in the `extractor_wrapper`? So, if I'm understanding correctly, in the test cases, the `pluck` value would be the raw string from the URL parameters.

File: panoptes_aggregation/tests/extractor_tests/test_question_extractor.py [hunk 9]
```diff
@@ -87,3 +87,37 @@
...
+            "uber_flag_digit": "4",
+            "is_gold_standard": "False"
+        }
+    }
+}
+
+single_pluck_keys = '{\
+    "gold_standard": "subject.metadata.is_gold_standard",\
+    "true_value": "subject.metadata.uber_flag_digit"\
+}'
```
Comment (CKrawczyk) [inline panoptes_aggregation/tests/extractor_tests/test_question_extractor.py:110]: This should not be a string given the way this is currently set up, pass it in as a dictionary.

**Review by CKrawczyk — COMMENTED** [review]
Changing this datatype should make the tests pass.

File: panoptes_aggregation/extractors/extractor_wrapper.py [hunk 10]
```diff
@@ -17,12 +18,20 @@ def decorator(func):
...
                 kwargs = argument.args.copy().to_dict()
                 if 'details' in kwargs:
                     kwargs['details'] = ast.literal_eval(kwargs['details'])
                 if 'tools' in kwargs:
                     kwargs['tools'] = ast.literal_eval(kwargs['tools'])
+                if 'pluck' in kwargs:
+                    pluck_key_dict = ast.literal_eval(kwargs['pluck'])
+
                 data = argument.get_json()
             else:
```
Comment (CKrawczyk) [inline panoptes_aggregation/extractors/extractor_wrapper.py:36]: Change to:

```python
else:
    data = argument
    pluck_key_dict = kwargs.pop('pluck', None)
```

File: panoptes_aggregation/tests/extractor_tests/test_question_extractor.py [hunk 11]
```diff
@@ -87,3 +87,37 @@
...
+    "pluck.gold_standard": "False",
+    "pluck.true_value": "4"
+}
+
+TestSinglePluck = ExtractorTest(extractors.question_extractor,
+    single_pluck_classification,
+    single_pluck_expected,
+    "Test pluck field functionality with a question extractor",
+    kwargs={'pluck': single_pluck_keys},
+    test_name='TestSinglePluck')
```
Comment (CKrawczyk) [inline panoptes_aggregation/tests/extractor_tests/test_question_extractor.py:123]: Small formatting change (looks slightly cleaner):

```python
TestSinglePluck = ExtractorTest(
    extractors.question_extractor,
    single_pluck_classification,
    single_pluck_expected,
    "Test pluck field functionality with a question extractor",
    kwargs={'pluck': single_pluck_keys},
    test_name='TestSinglePluck'
)
```

**Review by CKrawczyk — CHANGES_REQUESTED** [review]
I just tried these changes locally and they worked as expected.  Once these are in place it should be good to go.

**Review by CKrawczyk — APPROVED** [review]
🎉 All looks good.

Seems Coveralls is still down, if it is not back up by tomorrow I will force a merge, otherwise I will leave clicking the big green button to you.
